### PR TITLE
Partial #581: Add unit test for plugin load failures

### PR DESCRIPTION
**Pull request by lgolding**
When I introduced the plugin model for converters (PR #580), @michaelcfanning had a couple of suggestions, which I captured in Issue #581. This PR addresses the first of them: we add a unit test demonstrating that when the plugin assembly cannot be loaded, the `ToolFormatConverter` throws the expected exception.

Michael actually suggested that I add tests for `IOException`, `BadImageFormatException`, and `UnauthorizedAccessException`, but as you'll see, one test suffices.

To make this scenario testable, we again employ the "no-mocks" technique of injecting into the `ToolFormatConverter` a delegate that loads the plugin assembly. By default, that delegate is `Assembly.LoadFile`, but test code can supply a different delegate which (for example) throws an exception.

Also:
- Two unit tests in `ToolFormatConverterTests` were using XUnit assertions. Change them to use FluentAssertions.

### CODE REVIEW

File: src/Sarif.Converters/PluginConverterFactory.cs [hunk 1]
```diff
@@ -32,7 +39,7 @@ public override ToolFileConverterBase CreateConverterCore(string toolFormat)
                 throw new ArgumentException(message, nameof(this.pluginAssemblyPath));
             }
 
-            Assembly pluginAssembly = Assembly.LoadFile(this.pluginAssemblyPath);
+            Assembly pluginAssembly = this.assemblyLoadFileDelegate(this.pluginAssemblyPath);
```
Comment (jeff-bacon) [inline src/Sarif.Converters/PluginConverterFactory.cs]: >this. [](start = 68, length = 5)

Should the 'this' keyword be present outside of the constructor for this variable?

File: src/Sarif.Converters/ToolFormatConverter.cs [hunk 2]
```diff
@@ -100,12 +107,12 @@ public class ToolFormatConverter
...
         // factory capable of creating it.
         // This method is internal, rather than private, for test purposes.
-        internal static ConverterFactory CreateConverterFactory(string pluginAssemblyPath)
+        internal ConverterFactory CreateConverterFactory(string pluginAssemblyPath)
         {
             ConverterFactory factory = new BuiltInConverterFactory();
             if (!string.IsNullOrWhiteSpace(pluginAssemblyPath))
             {
-                factory = new PluginConverterFactory(pluginAssemblyPath)
+                factory = new PluginConverterFactory(pluginAssemblyPath, this.assemblyLoadFileDelegate)
```
Comment (jeff-bacon) [inline src/Sarif.Converters/ToolFormatConverter.cs]: >this. [](start = 73, length = 5)

Should the 'this' keyword be present outside of the constructor for this variable?

File: src/Sarif.Converters.UnitTests/ToolFormatConverterTests.cs [hunk 3]
```diff
@@ -144,6 +145,37 @@ public void ToolFormatConverter_FailsIfPluginAssemblyDoesNotExist()
...
+                string inputFilePath = tempDir.Write("input.txt", string.Empty);
+                string outputFilePath = tempDir.Combine("output.txt");
+
+                // Unlike all the other tests, which default-construct the converter, this test
+                // provides the converter with a delegate which simulates the failure to load the
+                // plugin assembly.
+                const string Message = "Something went dreadfully wrong.";
+                AssemblyLoadFileDelegate assemblyLoadFileDelegate = path => throw new BadImageFormatException(Message);
+
+                ToolFormatConverter converter = new ToolFormatConverter(assemblyLoadFileDelegate);
```
Comment (jeff-bacon) [inline src/Sarif.Converters.UnitTests/ToolFormatConverterTests.cs]: >ToolFormatConverter [](start = 16, length = 19)

`var` can be used here

File: src/Sarif.Converters/PluginConverterFactory.cs (hunk 1 above)
Comment (lgolding) [reply src/Sarif.Converters/PluginConverterFactory.cs]: In brief, yes. Our convention is to use `this.` for instance fields (check with Michael to confirm). The SARIF SDK code base does not follow this convention consistently.

File: src/Sarif.Converters/ToolFormatConverter.cs (hunk 2 above)
Comment (lgolding) [reply src/Sarif.Converters/ToolFormatConverter.cs]: See my reply above.

File: src/Sarif.Converters.UnitTests/ToolFormatConverterTests.cs (hunk 3 above)
Comment (lgolding) [reply src/Sarif.Converters.UnitTests/ToolFormatConverterTests.cs]: Fixed.

File: src/Sarif.Converters/PluginConverterFactory.cs (hunk 1 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PluginConverterFactory.cs]: Ah.  Seems like a good replacement for `_`.  

---
In reply to: [152131690](https://github.com/Microsoft/sarif-sdk/pull/685#discussion_r152131690) [](ancestors = 152131690)

**Pull request merged**